toolkit/tools/pkg/imagecustomizerlib/imagehistory.go

```diff
@@ -0,0 +1,245 @@
+	currentImageHistory := ImageHistory{
+		BuildTime:   buildTime,
+		ToolVersion: toolVersion,
+		ImageUuid:   imageUuid,
+		Config:      *configCopy,
+	}
+	allImageHistory = append(allImageHistory, currentImageHistory)
+
+	jsonBytes, err := json.MarshalIndent(allImageHistory, "", " ")
+	if err != nil {
+		return fmt.Errorf("failed to marshal image history: %w", err)
+	}
+
+	err = file.Write(string(jsonBytes), imageHistoryFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to write image history to file: %w", err)
+	}
+
+	return nil
+}
+
+func deepCopyConfig(config *imagecustomizerapi.Config) (*imagecustomizerapi.Config, error) {
+	configCopy := &imagecustomizerapi.Config{}
+	data, err := json.Marshal(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal config: %w", err)
+	}
+
+	err = json.Unmarshal(data, configCopy)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal config: %w", err)
+	}
+
+	return configCopy, nil
+}
+
+func modifyConfig(configCopy *imagecustomizerapi.Config, baseConfigPath string) error {
+	var err error
+	redactedString := "[redacted]"
+
+	err = populateScriptsList(configCopy.Scripts, baseConfigPath)
+	if err != nil {
+		return fmt.Errorf("failed to populate scripts list: %w", err)
+	}
+
+	err = populateAdditionalFiles(configCopy.OS.AdditionalFiles, baseConfigPath)
+	if err != nil {
+		return fmt.Errorf("failed to populate additional files: %w", err)
+	}
+
+	err = populateAdditionalDirs(configCopy.OS.AdditionalDirs, baseConfigPath)
+	if err != nil {
+		return fmt.Errorf("failed to populate additional dirs: %w", err)
+	}
+
+	err = redactSshPublicKeys(configCopy.OS.Users, redactedString)
+	if err != nil {
+		return fmt.Errorf("failed to redact ssh public keys: %w", err)
+	}
+	return nil
+}
+
+func populateAdditionalDirs(configAdditionalDirs imagecustomizerapi.DirConfigList, baseConfigPath string) error {
+	for i := range configAdditionalDirs {
+		hashes := make(map[string]string)
+		sourcePath := configAdditionalDirs[i].Source
+		dirPath := file.GetAbsPathWithBase(baseConfigPath, sourcePath)
+
+		addFileHashToMap := func(path string, d os.DirEntry, err error) error {
+			if err != nil {
+				return err
+			}
+
+			if d.IsDir() {
+				return nil
+			}
+
+			relPath, err := filepath.Rel(dirPath, path)
+			if err != nil {
+				return fmt.Errorf("error computing relative path for %s: %w", path, err)
+			}
+
+			hash, err := generateSHA256(path)
+			if err != nil {
+				return err
+			}
+
+			hashes[relPath] = hash
+			return nil
+		}
+
+		err := filepath.WalkDir(dirPath, addFileHashToMap)
+		if err != nil {
+			return fmt.Errorf("error walking directory %s: %w", dirPath, err)
+		}
+		configAdditionalDirs[i].SHA256HashMap = hashes
+	}
+	return nil
+}
+
+func populateAdditionalFiles(configAdditionalFiles imagecustomizerapi.AdditionalFileList, baseConfigPath string) error {
+	for i := range configAdditionalFiles {
+		if configAdditionalFiles[i].Source == "" {
```

Comment:
Can `Source` be an empty string?

Reply:
Yes. When content is provided, source is empty

Reply:
I flipped the logic to remove the use of continue